src/GafferScene/MeshSegments.cpp

```diff
@@ -0,0 +1,277 @@
+}
+
+const Gaffer::StringPlug *MeshSegments::indexPrimVarPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+bool MeshSegments::affectsProcessedObject( const Gaffer::Plug *input ) const
+{
+	return
+		ObjectProcessor::affectsProcessedObject( input ) ||
+		input == segmentsPrimVarPlug() ||
+		input == indexPrimVarPlug()
+	;
+}
+
+void MeshSegments::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ObjectProcessor::hashProcessedObject( path, context, h );
+
+	segmentsPrimVarPlug()->hash( h );
+	indexPrimVarPlug()->hash( h );
+}
+
+IECore::ConstObjectPtr MeshSegments::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, const IECore::Object *inputObject ) const
+{
+	const std::string &segmentsPrimVar = segmentsPrimVarPlug()->getValue();
+	const std::string &indexPrimVar = indexPrimVarPlug()->getValue();
+	const MeshPrimitive *mesh = runTimeCast<const MeshPrimitive>( inputObject );
+	if( !mesh || !segmentsPrimVar.size() )
+	{
+		return inputObject;
+	}
+
+	auto it = mesh->variables.find( indexPrimVar );
+	if( indexPrimVar != "" && it == mesh->variables.end() )
+	{
+		throw IECore::Exception( "No primitive variable named \"" + indexPrimVar + "\"" );
+	}
+
+	const std::vector<int> &verticesPerFace = mesh->verticesPerFace()->readable();
+	IntVectorDataPtr uniformSegmentsData = new IntVectorData();
+	if( indexPrimVar == "" || it->second.interpolation == PrimitiveVariable::Interpolation::Vertex || it->second.interpolation == PrimitiveVariable::Interpolation::Varying )
+	{
+		if( it->second.indices )
+		{
+			throw IECore::Exception( "Vertex primitive variable " + indexPrimVar + " must NOT be indexed in order to use as indexPrimVar, because specifying a vertex primitive variable as indexPrimVar means \"Use the mesh's vertex indices\", and should not be affected by the primitive variable at all." );
+		}
+		uniformSegmentsData = new IntVectorData();
+		segmentIndices( verticesPerFace, mesh->vertexIds()->readable(), mesh->variableSize( PrimitiveVariable::Interpolation::Vertex ), uniformSegmentsData->writable() );
+	}
+	else if( it->second.interpolation == PrimitiveVariable::Interpolation::FaceVarying )
+	{
+		if( !it->second.indices )
+		{
+			throw IECore::Exception( "FaceVarying primitive variable " + indexPrimVar + " must be indexed in order to use as indexPrimVar.  /todo : suggest using PrimitiveVariableWeld, once this node exists." );
+		}
+		uniformSegmentsData = new IntVectorData();
+		segmentIndices( verticesPerFace, it->second.indices->readable(), IECore::size( it->second.data.get() ), uniformSegmentsData->writable() );
+	}
+	else if( it->second.interpolation == PrimitiveVariable::Interpolation::Uniform )
+	{
+		if( !it->second.indices )
+		{
+			throw IECore::Exception( "Uniform primitive variable " + indexPrimVar + " must be indexed in order to use as indexPrimVar." );
+		}
+		uniformSegmentsData = it->second.indices;
```

Comment:
So, the final conclusion is that this is actually the right approach, right?  I haven't fully understood the explanation you gave... you said that copying a mesh creates copies of the primvars ( using copy-on-write proxies ), but it also somehow remembers which parts of the original mesh pointed to the same data, and makes them point to the same copied data - oh, I see this is what CopyContext does, I hadn't really looked at that before.  So, this means if I created a mesh by passing the same data to create P and Pref primvars, and then copied it, and then modified P, the copy-on-write mechanism would create a copy of the primvar so that we weren't modifying the original mesh, but we would be changing Pref as well?  And you're saying that we want to be able to do that kind of modification safely ... so does that mean that we have to ensure that we never share data between two different prim vars?  Seems a little counter-productive - how often do we actually modify a prim var in place in Gaffer, instead of replacing it with a fresh prim var?